### PR TITLE
Differential testing analytic engine: State file file dumper now dumps special files

### LINKED ISSUE #8990: Differential testing analytical engine: Dumper for special file store
System upgrade files are put in a special file store (`MerkleSpecialFiles`, a specialization of `MerkleLeaf`) and need to be dumped with the other files.

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

State file file dumper now dumps special files

* And for system files it tells you what file it is. 
* And you can ask for the report _without_ the file contents (much smaller file, suitable for putting into a spreadsheet.

**Related issue(s)**:

Fixes #8990 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (manual)

### CODE REVIEW

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java [hunk 1]
```diff
@@ -154,19 +233,28 @@ void reportSummary(
...
 
-        final var nMetadataFiles = fileSummary.typeOccurrences().get(VirtualBlobKey.Type.FILE_METADATA);
+        final var nMetadataFiles = fileSummary.typeOccurrences().getOrDefault(VirtualBlobKey.Type.FILE_METADATA, 0);
         final var nActiveDataFiles =
                 allFiles.values().stream().filter(HederaFile::isActive).count();
-        final var nActiveDataFilesWithMetadata =
-                allFiles.values().stream().filter(HederaFile::isActive).count();
+        final var nActiveDataFilesWithMetadata = allFiles.values().stream()
+                .filter(HederaFile::isActive)
+                .filter(hf -> null != hf.metadata())
```
Comment (mhess-swl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:245]:
```suggestion
                .filter(hf -> hf.metadata() != null)
```
Personally I'd flip these to make it more readable, but that's a completely subjective suggestion

**Review by mhess-swl — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java [hunk 2]
```diff
@@ -383,7 +488,49 @@ Pair<FileSummary, Map<Integer, HederaFile>> collectFiles(
...
+        }
+
+        final var typeMap = Map.of(VirtualBlobKey.Type.FILE_DATA, r.size());
+        final var fileSummary = new FileSummary(typeMap, Map.of(), 0);
+        return Pair.of(fileSummary, r);
+    }
+
+    /** Merge two (or more) maps.
+     *
+     * (Seems like it should be provided by the `Map` class, but maybe not.)
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:518]: Yes but why not a `Map.putAll` overload that takes a remapping function like `Map.merge`?  (Which is what I wanted here.)

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java [hunk 3]
```diff
@@ -114,32 +132,93 @@ void doit() {
...
-            return new HederaFile(contractId, contents, metadata);
+        static HederaFile of(@NonNull final FileStore fileStore, final int fileId, @NonNull final byte[] contents) {
+            return new HederaFile(fileStore, fileId, contents, null, SystemFileType.byId.get(fileId));
         }
 
         boolean isActive() {
-            return null != metadata && !metadata.isDeleted();
+            if (null != systemFileType) return true;
+            if (null != metadata) return !metadata.isDeleted();
+            return false;
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:221]: I get you, but many times - like here - I prefer the straight "ladder" of condition->return followed by the final return true/false.  And, TBH, a limitation of my brain is that your suggested conjunction is more difficult to follow that the simpler conditional followed by the bare return.

(Java really could use a null-coalescing or null-conditional operator like C#.)

Will leave this as-is.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java (hunk 1 above)
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:245]: So you don't like Yoda conditions?  It's an acquired taste to be sure.  (But you didn't complain about it up at L219-220?)

I just now tried `.filter(hf -> nonNull(hf.systemFileType())` (using `Objects::nonNull`) - I think I like it better and will use it.

(Actually what `Stream.filter` really needs is an overload `filter(Function<T,U> mapper, Predicate<U> predicate)` that does `e -> predicate(mapper(e))` so you could write, e.g., `filter(HederaFile::fileStore, isEqual(FileStore.SPECIAL))` - filtering on a condition of a field (or function) of the stream element but passing through the stream element.)

**Review by mhess-swl — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java (hunk 1 above)
Comment (mhess-swl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java]: This sparked an idea – a YouTube channel where Yoda reads source code as listeners fall asleep. You're welcome

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:519]: Map is already a rather large interface, and adding every useful option would make it unmanageable.  The assumption is that you will compose what's there (e.g. forEach with merge or replaceAll on the source map, then putAll to merge).
Using the Streams API is a reasonable approach, and mostly matches "recent" java practice (as well as being more "functional").

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java [hunk 4]
```diff
@@ -318,7 +424,7 @@ Pair<FileSummary, Map<Integer, HederaFile>> collectFiles(
 
         final var nType = new ConcurrentHashMap<VirtualBlobKey.Type, Integer>();
         final var nNullValues = new ConcurrentHashMap<VirtualBlobKey.Type, Integer>();
-        final var nNulLMetadataValues = new AtomicInteger();
+        final var nNullMetadataValues = new AtomicInteger();
```
Comment (jsync-swirlds) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:427]: What is the `n` prefix?  It's easy to miss if one doesn't read slowly.

Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpFilesSubcommand.java:427]: `n` == "number" (aka count) -

**Pull request merged**